### PR TITLE
Adding vm_storage_policy_profile resource for tag based placement policy management

### PR DESCRIPTION
**Pull request by sumitAgrawal007**
### Description
Adding VM_storage_policy_profile resource
CRUD operations for tag based placement policy rules

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
yes
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)
VM_storage_policy_profile acceptance test: https://gist.github.com/sumitAgrawal007/9db3c782a3a4af268f51ce0f9cdb30d5

tag category acceptance test
https://gist.github.com/sumitAgrawal007/91ff0b3fe9af063de0deca85242f51cd

### References
https://github.com/terraform-providers/terraform-provider-vsphere/issues/993

### CODE REVIEW

File: vsphere/internal/virtualdevice/virtual_machine_tag_rules_subresource.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	return map[string]*schema.Schema{
+		"tag_category": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "category",
+		},
+		"tags": {
+			Type:        schema.TypeList,
+			Required:    true,
+			Description: "tags",
```
Comment (shylasrinivas) [inline vsphere/internal/virtualdevice/virtual_machine_tag_rules_subresource.go:17]: description is cryptic..can you make it more descriptive? Applies to line 12 also..

File: vsphere/resource_vsphere_vm_storage_policy.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	types2 "github.com/vmware/govmomi/pbm/types"
+	"github.com/vmware/govmomi/vapi/tags"
+	"log"
+	"strings"
+)
+
+import (
+	"context"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
```
Comment (bill-rich) [inline vsphere/resource_vsphere_vm_storage_policy.go:16]: Combine imports to clean things up.

File: vsphere/resource_vsphere_vm_storage_policy.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+		},
+		"description": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Description of the storage policy",
+		},
+		"tag_rules": {
+			Type:        schema.TypeList,
+			Required:    true,
+			Description: "tag rules to filter datastores to be used for placement of VMs",
```
Comment (bill-rich) [inline vsphere/resource_vsphere_vm_storage_policy.go:36]: tag -> Tag

Comment (bill-rich) [inline vsphere/resource_vsphere_vm_storage_policy.go:36]: Punctuation on descriptions

File: vsphere/resource_vsphere_vm_storage_policy.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+		Create: resourceVmStoragePolicyCreate,
+		Read:   resourceVmStoragePolicyRead,
+		Update: resourceVmStoragePolicyUpdate,
+		Delete: resourceVmStoragePolicyDelete,
+		Schema: sch,
+	}
+}
+
+func resourceVmStoragePolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Print("[DEBUG] Beginning create")
```
Comment (bill-rich) [inline vsphere/resource_vsphere_vm_storage_policy.go:51]: Include resource name/type in log messages to make troubleshooting easier.

File: website/docs/r/vm_storage_policy.html.markdown [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  name = "policy1"
+  description = "description"
+
+  tag_rules {
+    tag_category = data.vsphere_tag_category.tag_category.name
+    tags = [data.vsphere_tag.tag1.name, data.vsphere_tag.tag2.name]
+    include_datastores_with_tags = true
+  }
+
+}
```
Comment (bill-rich) [inline website/docs/r/vm_storage_policy.html.markdown:54]: Run this through `terraform format` (or just line the = up manually)

File: vsphere/resource_vsphere_vm_storage_policy.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+		return fmt.Errorf("error while creating storage policy spec %s", err)
+	}
+
+	profileID, err := pbmClient.CreateProfile(context.Background(), *pbmCapabilityProfileCreateSpec)
+	if err != nil {
+		return fmt.Errorf("error while creating storage policy %s", err)
+	}
+
+	d.SetId(profileID.UniqueId)
+	log.Printf("[DEBUG] Storage policy Create complete with Id %s", profileID.UniqueId)
```
Comment (shylasrinivas) [inline vsphere/resource_vsphere_vm_storage_policy.go:120]: inconsistent use of case

File: website/docs/r/vm_storage_policy.html.markdown [hunk 7]
```diff
@@ -0,0 +1,68 @@
+---
+subcategory: "Storage"
+layout: "vsphere"
+page_title: "VMware vSphere: vm_storage_policy"
+sidebar_current: "docs-vsphere-resource-vm-storage-policy"
+description: |-
+  Provides CRUD operations on vm storage policy profiles. These policies help create tag based rules for placement of a 
+  VM on datastores. While placing a VM, Compatible Datastores can be filtered using these profiles.
```
Comment (shylasrinivas) [inline website/docs/r/vm_storage_policy.html.markdown:8]: Compatible Datastores -> compatible datastores

File: website/docs/r/vm_storage_policy.html.markdown [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+# vsphere\_vm\_storage\_policy
+
+The `vsphere_vm_storage_policy` resource can be used to create and manage storage 
+policies. Using this storage policy, tag based placement rules can be created to 
+place a VM on a particular tagged datastore.
+
+## Example Usage
+
+This example creates a storage policy with tag_rule having cat1 as tag_category and 
+tag1,tag2 as the tags. While creating a VM, this policy can be referenced to place
```
Comment (shylasrinivas) [inline website/docs/r/vm_storage_policy.html.markdown:20]: space between tag1, tag2

File: website/docs/r/vm_storage_policy.html.markdown [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the storage policy.
+* `description` - (Optional) Description of the storage policy. 
+* `tag_rules` - (Required) List of tag rules. The tag_category and tags to be associated in this profile.
```
Comment (shylasrinivas) [inline website/docs/r/vm_storage_policy.html.markdown:63]: suggest removing _ from tag_category in description

File: vsphere/resource_vsphere_vm_storage_policy.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+		for _, tagRule := range tagRules {
+
+			tagCategory := tagRule.(map[string]interface{})["tag_category"].(string)
+			_, err := tagCategoryByName(tagsManager, tagCategory)
+			if err != nil {
+				return fmt.Errorf("error while getting the tag %s %s", tagCategory, err)
+			}
+
+			tagValues := tagRule.(map[string]interface{})["tags"].([]interface{})
```
Comment (shylasrinivas) [inline vsphere/resource_vsphere_vm_storage_policy.go:231]: what happens when required params tag_category and tags are not provided?

Comment (sumitAgrawal007) [reply vsphere/resource_vsphere_vm_storage_policy.go:228]: tag_category is required, so terraform plan will throw an error.
Same with tags. Minimum 1 element is required.

**Pull request merged**